src/bgw/job_stat.c

```diff
@@ -397,17 +400,20 @@ calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failure
 			DirectFunctionCall2(timestamptz_pl_interval, TimestampTzGetDatum(last_finish), ival));
 		res_set = true;
 		ReleaseCurrentSubTransaction();
+		MemoryContextSwitchTo(oldctx);
+		CurrentResourceOwner = oldowner;
 	}
 	PG_CATCH();
 	{
+		RollbackAndReleaseCurrentSubTransaction();
 		MemoryContextSwitchTo(oldctx);
+		CurrentResourceOwner = oldowner;
 		ErrorData *errdata = CopyErrorData();
 		ereport(LOG,
 				(errcode(ERRCODE_INTERNAL_ERROR),
 				 errmsg("could not calculate next start on failure: resetting value"),
 				 errdetail("Error: %s.", errdata->message)));
 		FlushErrorState();
-		RollbackAndReleaseCurrentSubTransaction();
 	}
 	PG_END_TRY();
 	Assert(CurrentMemoryContext == oldctx);
```

Comment:
wouldn't it make sense to move 
```
	MemoryContextSwitchTo(oldctx);
		CurrentResourceOwner = oldowner;
```
from both try/catch branches to here - instead of this assert ?

it would make the `CopyErrorData` use a different mem context; but I think that doesn't make much difference

Reply:
This is following the pattern used in PostgreSQL to handle similar situations. Trying to "optimize" the code here runs 
the risk of introducing subtle errors. The comment on CopyErrorData says:

```
/*
 * CopyErrorData --- obtain a copy of the topmost error stack entry
 *
 * This is only for use in error handler code.  The data is copied into the
 * current memory context, so callers should always switch away from
 * ErrorContext first; otherwise it will be lost when FlushErrorState is done.
 */
```

Reply:
yes ; I know - but the code uses the results right away in the next line to create an error message...so it doesn't escape

it could be left as-is